server/src/main/java/org/elasticsearch/action/admin/indices/refresh/TransportUnpromotableShardRefreshAction.java

```diff
@@ -62,6 +72,46 @@ protected void unpromotableShardOperation(
         Task task,
         UnpromotableShardRefreshRequest request,
         ActionListener<ActionResponse.Empty> responseListener
+    ) {
+        if (clusterService.state().version() >= request.getClusterStateVersion()) {
+            doUnpromotableShardOperation(request, responseListener);
+        } else {
+            final Thread initialThread = Thread.currentThread();
+            ClusterStateObserver.waitForState(
+                clusterService,
+                clusterService.threadPool().getThreadContext(),
+                new ClusterStateObserver.Listener() {
+                    @Override
+                    public void onNewClusterState(ClusterState state) {
+                        if (initialThread == Thread.currentThread()) {
+                            doUnpromotableShardOperation(request, responseListener);
+                        } else {
+                            executor.execute(ActionRunnable.wrap(responseListener, l -> doUnpromotableShardOperation(request, l)));
+                        }
+                    }
+
+                    @Override
+                    public void onClusterServiceClose() {
+                        responseListener.onFailure(new NodeClosedException(clusterService.localNode()));
+                    }
+
+                    @Override
+                    public void onTimeout(TimeValue timeout) {
+                        assert false : "Timed out waiting for cluster state";
+                        responseListener.onFailure(new ElasticsearchException("Timed out waiting for cluster state"));
+                    }
+                },
+                state -> state.version() >= request.getClusterStateVersion(),
+                TimeValue.MAX_VALUE,
```

Comment:
The subsequent `shard.waitForPrimaryTermAndGeneration` does not have a timeout. So I chose to not timeout here as well. The caller side has a default timeout of 1min which should still kick in.